universal-application-tool-0.0.1/app/services/cloud/azure/BlobStorage.java

```diff
@@ -0,0 +1,241 @@
+package services.cloud.azure;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.azure.storage.blob.BlobClient;
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.BlobServiceClient;
+import com.azure.storage.blob.BlobServiceClientBuilder;
+import com.azure.storage.blob.models.BlobCorsRule;
+import com.azure.storage.blob.models.BlobServiceProperties;
+import com.azure.storage.blob.models.UserDelegationKey;
+import com.azure.storage.blob.sas.BlobSasPermission;
+import com.azure.storage.blob.sas.BlobServiceSasSignatureValues;
+import com.azure.storage.common.sas.SasProtocol;
+import com.typesafe.config.Config;
+import java.net.URL;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.util.List;
+import java.util.UUID;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import play.Environment;
+import services.cloud.StorageClient;
+import services.cloud.StorageServiceName;
+
+/**
+ * SimpleStorage provides methods to create federated links for users of CiviForm to upload and
+ * download files directly to and from AWS Simple Storage Service (S3).
+ */
+@Singleton
+public class BlobStorage implements StorageClient {
+
+  public static final String AZURE_STORAGE_ACCT_CONF_PATH = "azure.blob.account";
+  public static final String AZURE_CONTAINER_CONF_PATH = "azure.blob.container";
+  public static final String AZURE_REGION_CONF_PATH = "java.time.zoneid";
+  public static final Duration AZURE_SAS_TOKEN_DURATION = Duration.ofMinutes(10);
+  public static final Duration AZURE_USER_DELEGATION_KEY_DURATION = Duration.ofMinutes(60);
```

Comment:
question (blocking): what does the user delegation key duration do?

Reply:
The user delegation key is used to sign SAS tokens instead of using an account key and it uses "user" credentials (the term user is misleading here because it's actually using the application's Managed Identity credentials) to authenticate. The key needs to be rotated and this duration is the interval at which it's rotated. We weren't sure how long to make the interval. More info here: https://docs.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas

Reply:
Got it, thank you. Please document this in the code.

Reply:
done